shell/common/rasterizer.cc

```diff
@@ -802,28 +810,143 @@ static sk_sp<SkData> ScreenshotLayerTreeAsPicture(
+      nullptr,                      // thread merger
+      aiks_context.get()            // aiks context
+  );
+  canvas->Clear(DlColor::kTransparent());
+  frame->Raster(*tree, true, nullptr);
+  canvas->Flush();
+}
+
+#if IMPELLER_SUPPORTS_RENDERING
+Rasterizer::ScreenshotFormat ToScreenshotFormat(impeller::PixelFormat format) {
+  switch (format) {
+    case impeller::PixelFormat::kUnknown:
+    case impeller::PixelFormat::kA8UNormInt:
+    case impeller::PixelFormat::kR8UNormInt:
+    case impeller::PixelFormat::kR8G8UNormInt:
+    case impeller::PixelFormat::kR8G8B8A8UNormIntSRGB:
+    case impeller::PixelFormat::kB8G8R8A8UNormIntSRGB:
+    case impeller::PixelFormat::kB10G10R10XRSRGB:
+    case impeller::PixelFormat::kS8UInt:
+    case impeller::PixelFormat::kD24UnormS8Uint:
+    case impeller::PixelFormat::kD32FloatS8UInt:
+    case impeller::PixelFormat::kR32G32B32A32Float:
+    case impeller::PixelFormat::kB10G10R10XR:
+    case impeller::PixelFormat::kB10G10R10A10XR:
+      FML_DCHECK(false);
+      return Rasterizer::ScreenshotFormat::kUnknown;
+    case impeller::PixelFormat::kR8G8B8A8UNormInt:
+      return Rasterizer::ScreenshotFormat::kR8G8B8A8UNormInt;
+    case impeller::PixelFormat::kB8G8R8A8UNormInt:
+      return Rasterizer::ScreenshotFormat::kB8G8R8A8UNormInt;
+    case impeller::PixelFormat::kR16G16B16A16Float:
+      return Rasterizer::ScreenshotFormat::kR16G16B16A16Float;
+  }
+}
+
+static std::pair<sk_sp<SkData>, Rasterizer::ScreenshotFormat>
+ScreenshotLayerTreeAsImageImpeller(
+    const std::shared_ptr<impeller::AiksContext>& aiks_context,
+    flutter::LayerTree* tree,
+    flutter::CompositorContext& compositor_context,
+    bool compressed) {
+  if (compressed) {
+    FML_LOG(ERROR) << "Compressed screenshots not supported for Impeller";
+    return {nullptr, Rasterizer::ScreenshotFormat::kUnknown};
+  }
+
+  DisplayListBuilder builder(SkRect::MakeSize(
+      SkSize::Make(tree->frame_size().fWidth, tree->frame_size().fHeight)));
+
+  RenderFrameForScreenshot(compositor_context, &builder, tree, nullptr,
+                           aiks_context);
+
+  impeller::DlDispatcher dispatcher;
+  builder.Build()->Dispatch(dispatcher);
+  const auto& picture = dispatcher.EndRecordingAsPicture();
+  const auto& image = picture.ToImage(
+      *aiks_context,
+      impeller::ISize(tree->frame_size().fWidth, tree->frame_size().fHeight));
+  const auto& texture = image->GetTexture();
+  impeller::DeviceBufferDescriptor buffer_desc;
+  buffer_desc.storage_mode = impeller::StorageMode::kHostVisible;
+  buffer_desc.size =
+      texture->GetTextureDescriptor().GetByteSizeOfBaseMipLevel();
+  auto impeller_context = aiks_context->GetContext();
+  auto buffer =
+      impeller_context->GetResourceAllocator()->CreateBuffer(buffer_desc);
+  auto command_buffer = impeller_context->CreateCommandBuffer();
+  command_buffer->SetLabel("BlitTextureToBuffer Command Buffer");
```

Comment:
This code is very similar to the code we have in `vulkan_screenshotter.cc`.  Despite the name, it's actually doing this.  It would be nice if we could share this.  PTAL.

Reply:
There's a few places we have code a little like this, and trying to clean all of them up was getting to be pretty invasive. I think it's worth doing but didn't want to block the rest of this fix on it.